### PR TITLE
Add PyTorch Lightning experiment adapter (TorchExperiment)

### LINKED ISSUE #195: `torch` / `lightning` integration
It would be nice to have integration with basic `torch` and `lightning` tuning workflows, to allow autoML style tuning.

This would require a `BaseExperiment` descendant class `TorchExperiment` which takes a `DataLoader`, a `LightningModule`, and a `Trainer`, performs a training run, and returns the validation loss/score.

Since tuning is not API preserving in `torch` / `lightning`, I would suggest an additional function that produces parameters for a tuned network, or the initialized tuned network with said parameters.

Comments appreciated on the design.

### PR DESCRIPTION
**Pull request by amitsubhashchejara**
This pull request is to merge a new experiment `torch-lightning-experiment` for the issue `195`. The new class `TorchExperiment` is an experiment adapter for PyTorch Lightning experiments and is used to perform experiments using PyTorch Lightning modules. It allows for hyperparameter tuning and evaluation of the model's performance using specified metrics. 

The `TorchExperiment` class accepts a `LightningModule`, `DataModule`, `Trainer`, and an `objective_matric` with default value `val_loss`. The `_evaluate` function internally performs a training run and returns a score of the `objective_matric` as a `numpy.float64`.

Fixes #195

### CODE REVIEW

File: src/hyperactive/experiment/integrations/torch_lightning_experiment.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
+"""Experiment adapter for PyTorch Lightning experiments."""
+
+from hyperactive.base import BaseExperiment
+import torch
+import lightning as L
```
Comment (SimonBlanke) [inline src/hyperactive/experiment/integrations/torch_lightning_experiment.py:5]: I had to look this one up. First time I saw a package recommend a one uppercase letter import alias. We can leave it as is.

File: src/hyperactive/experiment/integrations/torch_lightning_experiment.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        Returns
+        -------
+        float
+            The value of the parameters as per evaluation.
+        dict
+            Additional metadata about the search.
+        """
+
+        model = self.lightning_module(**params)
+        self.trainer.fit(model, self.datamodule)
```
Comment (SimonBlanke) [inline src/hyperactive/experiment/integrations/torch_lightning_experiment.py:148]: Is the trainer designed to be fitted multiple/many times like this?

File: src/hyperactive/experiment/integrations/torch_lightning_experiment.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        # get validation results
+        if self.objective_metric not in self.trainer.callback_metrics:
+            raise ValueError(f"objective metric {self.objective_metric} not found in trainer callback metrics")
+        else:
+            val_results = self.trainer.callback_metrics.get(self.objective_metric, float('inf'))
+        metadata = {}
+        import numpy as np
```
Comment (SimonBlanke) [inline src/hyperactive/experiment/integrations/torch_lightning_experiment.py:155]: put this to the top of the file

File: src/hyperactive/experiment/integrations/torch_lightning_experiment.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+            Additional metadata about the search.
+        """
+
+        model = self.lightning_module(**params)
+        self.trainer.fit(model, self.datamodule)
+        # get validation results
+        if self.objective_metric not in self.trainer.callback_metrics:
+            raise ValueError(f"objective metric {self.objective_metric} not found in trainer callback metrics")
+        else:
+            val_results = self.trainer.callback_metrics.get(self.objective_metric, float('inf'))
```
Comment (SimonBlanke) [inline src/hyperactive/experiment/integrations/torch_lightning_experiment.py:153]: val_results -> val_result
It is one result, right?

**Review by SimonBlanke — CHANGES_REQUESTED** [review]
Deep-Learning frameworks often have very verbose logging behaviour per default. Since we run multiple trainings (in each iteration), this can get overwhelming in the command-line. Maybe we should address this.

Also answer the questions in the code and make necessary changes.

File: src/hyperactive/experiment/integrations/torch_lightning_experiment.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    ...     lightning_module=SimpleLightningModule,
+    ...     trainer=trainer,
+    ...     objective_metric="val_loss"
+    ... )
+    >>> 
+    >>> params = {"input_dim": 10, "hidden_dim": 16, "lr": 1e-3}
+    >>> 
+    >>> val_result, metadata = experiment._evaluate(params)
+    """
+
```
[the remark this answers is not in the document]
Comment (amitsubhashchejara) [reply src/hyperactive/experiment/integrations/torch_lightning_experiment.py:108]: Yeah, I was actually a bit confused about the usage of tags and the author credit. Can you help me with it.

**Comment by SimonBlanke** [comment]
@amitsubhashchejara Thanks for opening this PR. There is still some work to do, but I am sure you can do this.

**Comment by amitsubhashchejara** [comment]
Sure, I'll make the changes!

**Comment by amitsubhashchejara** [comment]
@SimonBlanke please approve the workflow!
